design-proposals/generalize-virt-stack/gen-virt-stack.md

```diff
@@ -0,0 +1,140 @@
+
+## Goals
+
+KubeVirt should be able to offer a choice to its users over which libvirt hypervisor-driver they want to use to create their VMI. The aim is to remove any hard-coded dependency of KubeVirt on qemu-kvm and have it interact solely with libvirt, while still being able to provide the same functionality for qemu-kvm based VMIs.
+
+Additionally, this design proposal intends to provide developers of other hypervisors with a flexible way of onboarding their own hypervisor to KubeVirt.
+
+## Non Goals
+
+Support all libvirt hypervisor-drivers in KubeVirt. As the developers of hypervisor `H` make its libvirt hypervisor-driver stable, they would add functionality to KubeVirt so that it can leverage hypervisor `H` to create VMIs.
+
+## Definition of Users
+
+This proposal is aimed at serving users who intend to use KubeVirt on a cluster where at least one node has a virtualization stack different from the default QEMU-KVM stack.
+
+## User Stories
+
+- A user trying to use KubeVirt on a cluster of machines that have a hypervisor-VMM pair that is not necessarily QEMU/KVM.
+- A regular user of libvirt with any of its supported hypervisor drivers who now wants to leverage the orchestration capability provided by KubeVirt. But the user does not want to abandon their hypervisor driver of choice.
+- A regular user of a hypervisor-VMM-specific orchestration capability to expand their use to a cluster with a diverse set of hypervisor-VMM pairs.
+
+## Repos
+
+- KubeVirt
+- Libvirt
+
+# Design
+
+## API Changes
+
+Addition of a `vmi.hypervisor` field. Example of how a user could request a specific hypervisor as the underlying libvirt hypervisor-driver through the VMI spec:
+
+```yaml
+spec:
+  hypervisor: cloud-hypervisor
+```
+
+- Introduction of a new field `hypervisor` in `VirtualMachineInstanceSpec`
+- By default if no `hypervisor` is provided, it would default to `qemu-kvm`.
+- The set of acceptable values for the `hypervisor` field would be the set of all hypervisor-drivers that libvirt supports.
+
+### Supported feature check
+
+Maintain a list of which VMI features are supported by different hypervisor-drivers. If the KubeVirt API user requests a feature for a VMI that is not supported by the requested hypervisor-driver, then the request for creation of VMI should be rejected.
+
+
+## Generalization of KubeVirt components
+
+### VMI spec to virt-launcher pod spec by Virt-Controller
+
+Conversion of the VMI spec to `virt-launcher` pod spec needs to take into account the `vmi.hypervisor` field. The value of this field would affect the following:
+
+- `virt-launcher` pod image should be specific to the `vmi.hypervisor`.
+
+- Hypervisor resource needed by the `virt-launcher` pod. For instance, for a VMI with `hypervisor=qemu-kvm`, the corresponding virt-launcher pod requires the resource `devices.kubevirt.io/kvm: 1`.
```

Comment:
Please, take emulation also into account

Reply:
Can you please expand on your comment?